src/main/java/bio/terra/service/filedata/azure/util/BlobContainerClientFactory.java

```diff
@@ -0,0 +1,187 @@
+      return createReadOnlySasUrlForBlobUsingUserDelegatedSas(blobName);
+    }
+
+    throw new RuntimeException("Failed to generate SAS token for blob. Invalid strategy set.");
+  }
+
+  private synchronized String createReadOnlySasUrlForBlobUsingUserDelegatedSas(String blobName) {
+
+    if (delegationKey == null
+        || delegationKey.getSignedExpiry().isBefore(OffsetDateTime.now(ZoneOffset.UTC))) {
+      OffsetDateTime keyExpiry = OffsetDateTime.now(ZoneOffset.UTC).plusDays(EXPIRATION_DAYS);
+      delegationKey = blobServiceClient.getUserDelegationKey(null, keyExpiry);
+    }
+
+    BlobServiceSasSignatureValues sasSignatureValues = createSasSignatureValues();
+
+    BlobClient blobClient = blobContainerClient.getBlobClient(blobName);
+    String sasToken = blobClient.generateUserDelegationSas(sasSignatureValues, delegationKey);
+
+    return String.format(
+        "%s/%s?%s", blobContainerClient.getBlobContainerUrl(), blobClient.getBlobName(), sasToken);
+  }
+
+  private String createReadOnlySasUrlForBlobUsingContainerSas(String blobName) {
+    return String.format(
+        "%s/%s?%s",
+        blobContainerClient.getBlobContainerUrl(),
+        blobName,
+        blobContainerSasTokenCreds.getSignature());
+  }
+
+  private String createReadOnlySasUrlForBlobUsingClient(String blobName) {
+
+    BlobServiceSasSignatureValues sasSignatureValues = createSasSignatureValues();
+
+    BlobClient blobClient = blobContainerClient.getBlobClient(blobName);
+    String sasToken = blobClient.generateSas(sasSignatureValues);
+
+    return String.format(
+        "%s/%s?%s", blobContainerClient.getBlobContainerUrl(), blobClient.getBlobName(), sasToken);
+  }
+
+  private BlobServiceSasSignatureValues createSasSignatureValues() {
+
+    BlobSasPermission permissions = new BlobSasPermission().setReadPermission(true);
+
+    OffsetDateTime expiryTime = OffsetDateTime.now().plusDays(1);
+    SasProtocol sasProtocol = SasProtocol.HTTPS_ONLY;
+
+    // build the token
+    return new BlobServiceSasSignatureValues(expiryTime, permissions).setProtocol(sasProtocol);
+  }
+
+  private BlobServiceClient createBlobServiceClientUsingSharedKey(
+      String accountName, String accountKey) {
+    return new BlobServiceClientBuilder()
+        .credential(new StorageSharedKeyCredential(accountName, accountKey))
+        .httpClient(httpClient)
+        .endpoint(String.format(Locale.ROOT, "https://%s.blob.core.windows.net", accountName))
+        .buildClient();
+  }
+
+  private BlobServiceClient createBlobServiceClientUsingTokenCredentials(
+      String accountName, TokenCredential credentials) {
+    return new BlobServiceClientBuilder()
+        .credential(credentials)
+        .httpClient(httpClient)
+        .endpoint(String.format(Locale.ROOT, "https://%s.blob.core.windows.net", accountName))
+        .buildClient();
+  }
+
+  public enum SasTokenGeneratorStrategy {
+    SharedKey,
+    ContainerSasToken,
+    UserDelegatedKey
```

Comment:
Enum members are actually instances of an enum, and they are final, so they are effectively static final objects. Therefore the java convention is to write them using all uppercase names.
```suggestion
    SHARED_KEY,
    CONTAINER_SAS_TOKEN,
    USER_DELEGATED_KEY
```

Reply:
Will fix. Also, I am noticing several style issues which are typically enforced/detected by configuration in the IDE or the linter. Is this any documentation in TDR, I can follow to configure my environment with the expected style checks?

Reply:
Do you have examples of one that you think could be caught by linters?  I could certainly see if I could enable that check with checkstyle or spotbugs config

Reply:
Yeah, we are (now) using spotless with the google style rules. I believe we do run findbugs too. Maybe we need to fiddle with the settings more to get these issues flagged, if possible. I think IntelliJ has settings for these too, but I think they're inspections and not editor settings that are easily shared. 

I agree that it would be best to have our IDEs and CI report these errors, rather than using the PR process to address them!